File: go/day01/task2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"unicode"
)

var digitMap = map[string]string{
	"one":   "1",
	"two":   "2",
	"three": "3",
	"four":  "4",
	"five":  "5",
	"six":   "6",
	"seven": "7",
	"eight": "8",
	"nine":  "9",
}

func matchDigit(s string) string {
	for word, digit := range digitMap {
		if strings.HasPrefix(s, word) {
			return digit
		}
	}
	return ""
}

func findDigits(s string) (string, string) {
	firstDigit, lastDigit := "", ""
	for pos := 0; pos < len(s); pos++ {
		ch := rune(s[pos])
		if unicode.IsLetter(ch) {
			digit := matchDigit(s[pos:])
			if digit != "" {
				if firstDigit == "" {
					firstDigit = digit
				}
				lastDigit = digit
			}
		} else if unicode.IsDigit(ch) {
			if firstDigit == "" {
				firstDigit = string(ch)
			}
			lastDigit = string(ch)
		}
	}
	return firstDigit, lastDigit
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	sum := 0
	for scanner.Scan() {
		line := scanner.Text()
		firstDigit, lastDigit := findDigits(strings.ToLower(line))
		if firstDigit != "" && lastDigit != "" {
			number, _ := strconv.Atoi(firstDigit + lastDigit)
			sum += number
		}
	}
	if scanner.Err() != nil {
		fmt.Fprintln(os.Stderr, "reading standard input:", scanner.Err())
	}
	fmt.Println(sum)
}

File: polars/day01/task2.py

import sys
import polars as pl

rows = []
for line in sys.stdin:
    rows.append(line.strip())
df = pl.DataFrame([rows], schema=["input"])

numbers_map = {}
for i, num in enumerate(["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]):
    numbers_map[num] = num + str(i+1) + num

df = df.with_columns(replaced=pl.col("input"))
for num, digit in numbers_map.items():
    df = df.with_columns(replaced=pl.col("replaced").str.replace_all(num, digit, literal=True))

answer = df.with_columns(
    digits=pl.col("replaced").str.extract_all(r"(\d)")
).with_columns(
    first_digit=pl.col("digits").list.first().cast(pl.Int64),
    last_digit=pl.col("digits").list.last().cast(pl.Int64),
).with_columns(
    sum=pl.col("first_digit")*10 + pl.col("last_digit"),
)["sum"].sum()

print(answer)